src/Indicator.vala

```diff
@@ -82,15 +81,25 @@ public class DateTime.Indicator : Wingpanel.Indicator {
         return main_grid;
     }
 
+    private void create_event_grid () {
+        event_grid = new Gtk.Grid ();
+        event_grid.orientation = Gtk.Orientation.VERTICAL;
+        main_grid.attach (event_grid, 0, 1);
+    }
+
     private void update_events_model (E.Source source, Gee.Collection<E.CalComponent> events) {
         Idle.add (update_events);
```

Comment:
To prevent this being called twice in different signal handlers, you can create a private uint variable to hold the source id and check whether it is non-zero before creating the Idle.  If it is, remove the pending source and create a new one so that only the last call actually does the update.  Set the source id back to zero at the end of the idle closure.

To DRY the code the checking and creation of the Idle should be in a separate function.  I am not entirely sure the Idle is necessary.  I removed it and connected all the signal handlers that use it with `connect_after` and did not notice a regression.  However, it may be safer to keep it.

I thought that update_events being called concurrently in different threads might have been causing the problem but unfortunately preventing that did not fix it.  I think it is just a bug/feature of Grid.